Astronomy/View Controllers/PhotosCollectionViewController.swift

```diff
@@ -60,12 +66,59 @@ class PhotosCollectionViewController: UIViewController, UICollectionViewDataSour
         return UIEdgeInsets(top: 0, left: 10.0, bottom: 0, right: 10.0)
     }
     
+    
+    // Cancel
+    
+    func collectionView(_ collectionView: UICollectionView, didEndDisplaying cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
+        
+        let photoRefence = photoReferences[indexPath.item]
+        operations[photoRefence.id]?.cancel()
+    }
+    
+    
+    
     // MARK: - Private
     
     private func loadImage(forCell cell: ImageCollectionViewCell, forItemAt indexPath: IndexPath) {
         
         // let photoReference = photoReferences[indexPath.item]
         
+        let photoReference = photoReferences[indexPath.item]
+        
+        if let cacheData = cache.value(key: photoReference.id) {
+            cell.imageView.image = UIImage(data: cacheData)
+            return
+        } else {
+            
+            let fetchPhoto = FetchPhotoOperation(photoReference: photoReference)
+            
+            let cachePhoto = BlockOperation {
+                if let imageData = fetchPhoto.imageData {
+                    self.cache.cache(value: imageData, for: photoReference.id)
+                }
+            }
+        
+            let cellData = BlockOperation {
+                guard let imageData = fetchPhoto.imageData else { return }
+                DispatchQueue.main.async {
+                    if self.collectionView.indexPath(for: cell) == indexPath {
```

Comment:
This method is not reliable. I would suggest instead putting an additional property on the cell that has the photo reference id and checking that to determine if it is the same cell. Notice that some of your images don't load currently.